### PR TITLE
Add table azure_search_service closes #240

### LINKED ISSUE #240: Add table azure_search_service
**References**
Add any related links that will help us understand the resource, including vendor documentation, related GitHub issues, and Go SDK documentation.

### PR DESCRIPTION
**Pull request by ParthaI**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/azure_search_service []

PRETEST: tests/azure_search_service

TEST: tests/azure_search_service
Running terraform
data.azurerm_client_config.current: Refreshing state...
data.null_data_source.resource: Refreshing state...
azurerm_resource_group.named_test_resource: Creating...
azurerm_resource_group.named_test_resource: Creation complete after 3s [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest99778]
azurerm_search_service.named_test_resource: Creating...
azurerm_search_service.named_test_resource: Still creating... [10s elapsed]
azurerm_search_service.named_test_resource: Creation complete after 16s [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest99778/providers/Microsoft.Search/searchServices/turbottest99778]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

Outputs:

location = eastus
resource_aka = azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest99778/providers/Microsoft.Search/searchServices/turbottest99778
resource_aka_lower = azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourcegroups/turbottest99778/providers/microsoft.search/searchservices/turbottest99778
resource_id = /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest99778/providers/Microsoft.Search/searchServices/turbottest99778
resource_name = turbottest99778
subscription_id = d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8

Running SQL query: test-get-query.sql
[
  {
    "id": "/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest99778/providers/Microsoft.Search/searchServices/turbottest99778",
    "name": "turbottest99778",
    "provisioning_state": "succeeded",
    "region": "east us",
    "resource_group": "turbottest99778",
    "sku_name": "standard",
    "status": "running",
    "subscription_id": "d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8",
    "tags_src": {},
    "type": "Microsoft.Search/searchServices"
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "name": "turbottest99778",
    "network_rule_set": {
      "ipRules": []
    },
    "partition_count": 1,
    "replica_count": 1
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "id": "/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest99778/providers/Microsoft.Search/searchServices/turbottest99778",
    "name": "turbottest99778",
    "region": "east us"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest99778/providers/Microsoft.Search/searchServices/turbottest99778",
      "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourcegroups/turbottest99778/providers/microsoft.search/searchservices/turbottest99778"
    ],
    "name": "turbottest99778",
    "title": "turbottest99778"
  }
]
✔ PASSED

POSTTEST: tests/azure_search_service

TEARDOWN: tests/azure_search_service

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
### Basic Info

```sql
select
  name,
  id,
  type,
  provisioning_state,
  status,
  sku_name,
  replica_count
from
  azure_search_service;
```

```
+---------------+--------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+------------------
| name          | id                                                                                                                                   | type                            | provisioning_stat
+---------------+--------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+------------------
| my-service-23 | /subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbot_rg/providers/Microsoft.Search/searchServices/my-service-23 | Microsoft.Search/searchServices | succeeded        
+---------------+--------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+------------------

```
```
> select
  name,
  id,
  public_network_access
from
  azure_search_service
where
  public_network_access = 'Enabled';
+--------+-----------------------------------------------------------------------------------------------------------------------------------+-----------------------+
| name   | id                                                                                                                                | public_network_access |
+--------+-----------------------------------------------------------------------------------------------------------------------------------+-----------------------+
| test43 | /subscriptions/000000-f95f-00bbb5-000000/resourceGroups/hipaa-hitrust/providers/Microsoft.Search/searchServices/test43 | Enabled               |
+--------+-----------------------------------------------------------------------------------------------------------------------------------+-----------------------+
```
</details>

### CODE REVIEW

File: azure-test/tests/azure_search_service/test-turbot-expected.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[
+	{
+		"akas": [
+			"{{ output.resource_aka.value }}",
+			"{{ output.resource_aka_lower.value }}"
+		],
+		"name": "{{ resourceName }}",
+		"title": "{{ resourceName }}"
+	}
+]
```
Comment (khushboo9024) [inline azure-test/tests/azure_search_service/test-turbot-expected.json:10]: Fix the indentation.

File: azure/table_azure_search_service.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+			{
+				Name:        "id",
+				Type:        proto.ColumnType_STRING,
+				Description: "Fully qualified resource ID for the resource.",
+				Transform:   transform.FromGo(),
+			},
+			{
+				Name:        "provisioning_state",
+				Type:        proto.ColumnType_STRING,
+				Description: "The state of the last provisioning operation performed on the search service. Provisioning is an intermediate state that occurs while service capacity is being established. After capacity is set up, provisioningState changes to either 'succeeded' or 'failed'. Client applications can poll provisioning status (the recommended polling interval is from 30 seconds to one minute) by using the Get Search Service operation to see when an operation is completed. If you are using the free service, this value tends to come back as 'succeeded' directly in the call to Create search service. This is because the free service uses capacity that is already set up. Possible values include: 'Succeeded', 'Provisioning', 'Failed'.",
```
Comment (khushboo9024) [inline azure/table_azure_search_service.go:41]:
```suggestion
				Description: "The state of the last provisioning operation performed on the search service.",
```

File: azure/table_azure_search_service.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+			},
+			{
+				Name:        "status",
+				Type:        proto.ColumnType_STRING,
+				Description: "The status of the search service. Possible values include: 'running': The search service is running and no provisioning operations are underway. 'provisioning': The search service is being provisioned or scaled up or down. 'deleting': The search service is being deleted. 'degraded': The search service is degraded. This can occur when the underlying search units are not healthy. The search service is most likely operational, but performance might be slow and some requests might be dropped. 'disabled': The search service is disabled. In this state, the service will reject all API requests. 'error': The search service is in an error state. If your service is in the degraded, disabled, or error states, it means the Azure Cognitive Search team is actively investigating the underlying issue. Dedicated services in these states are still chargeable based on the number of search units provisioned. Possible values include: 'ServiceStatusRunning', 'ServiceStatusProvisioning', 'ServiceStatusDeleting', 'ServiceStatusDegraded', 'ServiceStatusDisabled', 'ServiceStatusError'",
```
Comment (khushboo9024) [inline azure/table_azure_search_service.go:47]:
```suggestion
				Description: "The status of the search service. Possible values include: 'running', deleting', 
 'provisioning', 'degraded', 'disabled', 'error' etc.
```

File: azure/table_azure_search_service.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+				Transform:   transform.FromField("ServiceProperties.Status"),
+			},
+			{
+				Name:        "status_details",
+				Type:        proto.ColumnType_STRING,
+				Description: "The status of the search service. Possible values include: 'runniThe details of the search service status.",
```
Comment (khushboo9024) [inline azure/table_azure_search_service.go:53]:
```suggestion
				Description: "The details of the search service status.",
```

File: azure/table_azure_search_service.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+				Description: ColumnDescriptionTags,
+				Type:        proto.ColumnType_JSON,
+			},
+			{
+				Name:        "akas",
+				Description: ColumnDescriptionAkas,
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("ID").Transform(idToAkas),
+			},
+			{
```
Comment (khushboo9024) [inline azure/table_azure_search_service.go:140]:
```suggestion
                        // Azure standard columns
			{
```

File: azure/table_azure_search_service.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+//// HYDRATE FUNCTIONS
+
+func getSearchService(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("getSearchService")
+
+	name := d.KeyColumnQuals["name"].GetStringValue()
+	resourceGroup := d.KeyColumnQuals["resource_group"].GetStringValue()
+
```
Comment (khushboo9024) [inline azure/table_azure_search_service.go:202]: Add check for empty name and resource_group column.

**Review by khushboo9024 — COMMENTED** [review]
Add diagnosticSettings details for search service.

File: azure/table_azure_search_service.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+				Transform:   transform.FromField("ServiceProperties.SharedPrivateLinkResources"),
+			},
+			{
+				Name:        "tags_src",
+				Type:        proto.ColumnType_JSON,
+				Description: "The resource tags.",
+				Transform:   transform.FromField("Tags"),
+			},
+
+			// Azure standard columns
```
Comment (khushboo9024) [inline azure/table_azure_search_service.go:122]:
```suggestion
			// Steampipe standard columns
```

**Review by khushboo9024 — APPROVED** [review]

File: azure/table_azure_search_service.go [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+			{
+				Name:        "shared_private_link_resources",
+				Type:        proto.ColumnType_JSON,
+				Description: "The list of shared private link resources managed by the azure cognitive search service.",
+				Transform:   transform.FromField("ServiceProperties.SharedPrivateLinkResources"),
+			},
+			{
+				Name:        "tags_src",
+				Type:        proto.ColumnType_JSON,
+				Description: "The resource tags.",
```
Comment (bigdatasourav) [inline azure/table_azure_search_service.go:126]: we do not need tags_src column, as Tags type is - map[string]*string

File: azure-test/tests/azure_search_service/test-get-expected.json [hunk 9]
```diff
@@ -0,0 +1,14 @@
+[
+  {
+    "id": "{{ output.resource_id.value }}",
+    "name": "{{ resourceName }}",
+    "provisioning_state": "succeeded",
+    "region": "east us",
```
Comment (bigdatasourav) [inline azure-test/tests/azure_search_service/test-get-expected.json:6]: we should minimize hard coded value in expected json. we can set output values and then use.

File: azure/table_azure_search_service.go [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+			{
+				Name:        "status_details",
+				Type:        proto.ColumnType_STRING,
+				Description: "The details of the search service status.",
+				Transform:   transform.FromField("ServiceProperties.StatusDetails"),
+			},
+			{
+				Name:        "type",
+				Type:        proto.ColumnType_STRING,
+				Description: "The type of the resource. E.g. 'Microsoft.Compute/virtualMachines' or 'Microsoft.Storage/storageAccounts'.",
```
Comment (bigdatasourav) [inline azure/table_azure_search_service.go:60]:
```suggestion
				Description: "The type of the resource.",
```

File: azure/table_azure_search_service.go [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+				Description: "A list of active diagnostic settings for the search service.",
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     listSearchServiceDiagnosticSettings,
+				Transform:   transform.FromValue(),
+			},
+			{
+				Name:        "identity",
+				Type:        proto.ColumnType_JSON,
+				Description: "The identity of the resource.",
+				Transform:   transform.FromField("Identity"),
```
Comment (bigdatasourav) [inline azure/table_azure_search_service.go:103]: do we need transform here.

File: azure-test/tests/azure_search_service/test-get-expected.json (hunk 9 above)
Comment (bigdatasourav) [reply azure-test/tests/azure_search_service/test-get-expected.json:6]: also remove empty value checks.

**Review by bigdatasourav — APPROVED** [review]
LGTM

**Review by rajlearner17 — APPROVED** [review]
LGTM

**Pull request merged**